### ISSUE TITLE
RFE: improved message / guidance for mixup of -eq with =

### ISSUE THREAD

**Issue by jcbollinger** [koalaman/shellcheck#2261]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
foo=
[[ "${foo}" -eq none ]]
```

#### Here's what shellcheck currently says:
```
Line 3	SC2154: none is referenced but not assigned.
```
#### Here's what I wanted or expected to see:

ShellCheck is right: bash will perform arithmetic expansion of the token `none`, which was intended to be interpreted as a string, not a variable name.  But the real issue here is that the wrong comparison operator was used, `-eq` instead of `=`.  That's *why* the arithmetic expansion is performed, as ShellCheck also recognizes.  ShellCheck could have been a lot more helpful for diagnosing that than in fact it was. At minimum, the guidance page for SC2154 could say something about how arithmetic expansion can be involved in an issue of this kind.

**Comment by TinCanTech** [comment]
The wiki is open to edits and improvements ..

**Comment by jcbollinger** [comment]
I'll consider having a go at the wiki, but the more I think about it, the more I wish ShellCheck would convey more contextual information in this case.  It is probably too much too ask that it should correctly diagnose the mixup of `-eq` with `=`, but I am inclined to think that it much more easily could simply have a separate error code for referenced but not assigned variables appearing in the context of an arithmetic expansion, or perhaps specifically for those variables that, by virtue of arithmetic context, are expanded despite not having the usual `$` telltale of a variable reference.

**Comment by TinCanTech** [comment]
Did you try `shellcheck` with this:
```
foo=
none="0"
[[ "${foo}" -eq none ]]
```

**Comment by jcbollinger** [comment]
Thanks, but I know how to fix the script.  I'm just here because ShellCheck's guidance in this case turned out to be far less helpful than usual.  Even after it told me I had an issue (in my original, larger script), I had trouble seeing it or recognizing its nature.

But thank goodness for ShellCheck all the same, as it at least spotted that there *was* an issue.  I could have wasted a lot more time trying to troubleshoot that without ShellCheck's help.

**Comment by TinCanTech** [comment]
Actually, the script I posted I would expect `shellcheck` to spot the discrepancy between using `-eq` with a string variable .. it was not meant as a _fix_.

**Comment by jcbollinger** [comment]
Oh.  Well, ShellCheck doesn't report any issues with that script.

**Comment by jcbollinger** [comment]
What's happening with the script is this, as best I can determine:

 - `[[` is a Bash keyword, as opposed to an external command or even a built-in one, so the evaluation of the associated expression is performed in the current shell environment, during expansion of the command.
 - *after* parameter expansion and quote removal, the expression is subject to arithmetic expansion.  It is at this point that, in the context of an `-eq` operator, the token `none` is interpreted as a variable reference.
   - when evaluated in this context, operands of `-eq` that do not expand to digit sequences are interpreted as 0
   - the secondary variable expansion and interpretation of non-numeric operands as 0 are indeed special behaviors of `[[`.  They are not documented for POSIX `[` and `test` commands, and they are not exhibited by bash's built-in versions of those or by the external versions of the GNU coreutils.  All of these latter reject `none` as an invalid operand for `-eq`, which ShellCheck also knows about and will diagnose where appropriate.
